relay-general/src/pii/attachments.rs

```diff
@@ -71,43 +65,283 @@ fn apply_regex_to_bytes(
+        }
+        if text_offset == re_match.end() {
+            encoded_end = encoded_offset;
+            break;
+        }
+    }
+
+    &mut all_encoded[encoded_start..encoded_end]
+}
+
+/// Traits to modify the strings in ways we need.
+trait StringMods: AsRef<[u8]> {
+    type Error;
+
+    /// Replace this string's contents by repeating the given character into it.
+    ///
+    /// # Panics
+    ///
+    /// The `fill_char` has to encode to the smallest encoding unit, otherwise this will
+    /// panic.  Using an ASCII replacement character is usually safe in most encodings.
+    fn fill_content(&mut self, fill_char: char);
+
+    /// Replace this string's contents with the given replacement string.
+    ///
+    /// If the replacement string encodes to a shorter byte-slice than the current string
+    /// any remaining space will be filled with the padding character.
+    ///
+    /// If the replacement string encodes to a longer byte-slice than the current string the
+    /// replacement string is truncated.  If this does not align with a character boundary
+    /// in the replacement string it is further trucated to the previous character boundary
+    /// and the remainder is filled with the padding char.
+    ///
+    /// # Panics
+    ///
+    /// The `padding` character has to encode to the smallest encoding unit, otherwise this
+    /// will panic.  Using an ASCII padding character is usually safe in most encodings.
+    fn swap_content(&mut self, replacement: &str, padding: char);
+
+    // /// Scrub a single regex match.
+    // ///
+    // /// This doesn't care if the match is a group or the entire match, the scrubbing
+    // /// behaviour is the same.
+    // fn scrub_match(&mut self, all_text: &str, re_match: Match, redaction: &Redaction) {
+
+    /// Apply a PII scrubbing redaction to this string slice.
+    fn apply_redaction(&mut self, redaction: &Redaction) {
+        const PADDING: char = 'x';
+        const MASK: char = '*';
+
+        match redaction {
+            Redaction::Default | Redaction::Remove => {
+                self.fill_content(PADDING);
+            }
+            Redaction::Mask => {
+                self.fill_content(MASK);
+            }
+            Redaction::Hash => {
+                let hashed = hash_value(self.as_ref());
+                self.swap_content(&hashed, PADDING);
+            }
+            Redaction::Replace(ref replace) => {
+                self.swap_content(replace.text.as_str(), PADDING);
             }
         }
-        Redaction::Replace(ref replace) => {
-            for (start, end) in matches {
-                replace_bytes_padded(
-                    replace.text.as_bytes(),
-                    &mut data[start..end],
-                    DEFAULT_PADDING,
-                );
+    }
+}
+
+impl StringMods for WStr {
+    type Error = Utf16Error;
+
+    fn fill_content(&mut self, fill_char: char) {
+        let size = std::mem::size_of::<u16>();
```

Comment:
I do this in a few functions.  However `std::mem::size_of` is a `const fn` so this is probably just a no-op?  In that case I'd prefer to write`std::mem::size_of::<u16>()` everywhere instead.  Unless I'm alone in this.